### PR TITLE
EBS Root Volume Termination

### PR DESCRIPTION
**Pull request by tioxy**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Enables to change EBS DeleteOnTermination by specifiying separated flags for *Root Volume* and *Additional Volumes* inside `InstanceGroupSpec`.

**Which issue(s) this PR fixes**:
Fixes #6373

**Special notes for your reviewer**:

> Further discussion about UX changed field names. Previous source code is [here](https://github.com/tioxy/kops/tree/ebs_termination_old)

We found better to split `deleteOnTermination` into different fields, as it enabled a deeper customization of InstanceGroups (originally was suggested only a single field in the structure).

Now we can finally close the **1.15** Milestone :smile:   
Also, a huge thanks to @joshbranham's work

**Does this PR introduce a user-facing change?**:
```release-note
Add capacity to retain EBS volumes after instance termination from InstanceGroups
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
---
kind: InstanceGroup
spec:
  rootVolumeDeleteOnTermination: false

---
kind: InstanceGroup
spec:
  volumes:
  - device: /dev/xvdd
    deleteOnTermination: false
```

### CODE REVIEW

**Comment by rifelpet** [comment]
/ok-to-test

Can you add a test for this? Perhaps just adding the field in [this manifest](https://github.com/kubernetes/kops/blob/master/tests/integration/update_cluster/complex/in-v1alpha2.yaml) and updating the terraform and cloudformation files in the same directory to match.

**Comment by joshbranham** [comment]
@tioxy have you tried building an instance group with and without this setting? I had some difficulty with an existing group and the `kops update cluster` diff showing the change to false, but maybe you handle that more elegantly than I had!

**Comment by tioxy** [comment]
Hi @joshbranham :grinning: 

One of the changes that probably affected your attempt was around `awstasks/launchconfiguration.go` with the method `Find()`. It will probably output mislead information if you do not change the actual Launch Configuration struct field.
```
    actual.RootVolumeTermination = b.Ebs.DeleteOnTermination
```

I made a step-by-step below running kops updates against different instance group volume settings as you suggested :+1: 
> I cleaned the ouput as much as I could to make it more clear

These are the manifests that I used to create a cluster with different volume settings:
```yaml
 kind: InstanceGroup
 metadata:
   name: master-us-east-1a
 spec:
   role: Master
   rootVolumeRetainOnTermination: true     # retain root
   volumes:
   - device: /dev/xvdd
     encrypted: true
     size: 20
     type: gp2
```

```yaml
 kind: InstanceGroup
 metadata:
   name: nodes
 spec:
   role: Node
   volumes:
   - device: /dev/xvdd
     encrypted: true
     retainOnTermination: true     # retain xvdd
     size: 20
     type: gp2
```

Then I ran `kops update cluster` (1st output):
```go
  LaunchConfiguration/master-us-east-1a.masters.kops.tioxy.com
  	BlockDeviceMappings 	[{"DeviceName":"/dev/xvdd","EbsDeleteOnTermination":true,"EbsEncrypted":true,"EbsVolumeIops":null,"EbsVolumeSize":20,"EbsVolumeType":"gp2","VirtualName":null}]
  	RootVolumeSize      	64
        // retain root
  	RootVolumeTermination	false
  	RootVolumeType      	gp2	

  LaunchConfiguration/nodes.kops.tioxy.com
        // retain xvdd
  	BlockDeviceMappings 	[{"DeviceName":"/dev/xvdd","EbsDeleteOnTermination":false,"EbsEncrypted":true,"EbsVolumeIops":null,"EbsVolumeSize":20,"EbsVolumeType":"gp2","VirtualName":null}]
  	RootVolumeSize      	128
  	RootVolumeTermination	true
  	RootVolumeType      	gp2      	
```

I edited the `nodes` InstanceGroup to add `rootVolumeRetainOnTermination: true`:

```yaml
 kind: InstanceGroup
 metadata:
   name: nodes
 spec:
   role: Node
   rootVolumeRetainOnTermination: true     # retaining root now
   volumes:
   - device: /dev/xvdd
     encrypted: true
     retainOnTermination: true
     size: 20
     type: gp2
```


And this the result after `kops update cluster` (2nd output):
```go
Will modify resources:
  LaunchConfiguration/nodes.kops.tioxy.com
        // retaining root now
  	RootVolumeTermination	 true -> false

Must specify --yes to apply changes
```

**Comment by tioxy** [comment]
@rifelpet Even after `kops delete cluster`, the retained volumes will be available in the AWS account (which I think is the intended behavior).

Should we be worried about these dangling volumes during integration tests (costs and unused resources)?

**Comment by tioxy** [comment]
Fixed tests 👍

**Comment by geojaz** [comment]
@tioxy I think this looks good, and I'm trying to give it a test over here for you, but I noticed that I'm getting a failure on verify-goimports (which `make ci` runs). I'll take a look at the rest of it, but could you take a look at that one pls?

**Comment by tioxy** [comment]
@geojaz I've just found out what was happening: an indentation problem in the generated `upup/models/bindata.go` file :sweat_smile:

To get it, I ran the following command:

```sh
goimports -v -d upup/models/bindata.go
```
[Full logs here](https://gist.githubusercontent.com/tioxy/661b5987e46056aedf58c2344ae7efde/raw/8a879827899ff372dc009b1604f70e95b393bf1e/goimports.out)

I found better to add `goimports` before running `verify-goimports` in `make ci`, as it will cause autogenerated files to be indented properly. Do you have any suggestions on it?

https://github.com/kubernetes/kops/pull/7865/files#diff-b67911656ef5d18c4ae36cb6741b7965R564

**Comment by geojaz** [comment]
@tioxy i think [this](https://github.com/kubernetes/kops/pull/7865/files#diff-b67911656ef5d18c4ae36cb6741b7965R564) is a useful change. Would you mind pulling that change into it's own PR? You can tag me on it and i'll give it a 👍 

as for the substance of your proposed change... I'm on-board with the end goal, which I understand to be something like: "make the volume retention policy easily configurable (for both the root volume and any additional volumes)" if that's wrong, please correct me :)

however, I ran through a couple test scenarios to check these config options out, and at least to me, the UX is a bit confusing. IMO, there are a couple issues with current implementation that i'm not sure how to resolve... 

1. (side note) root volume and additional volume config seem inconsistent. that is- root volume flags are at the top level under the IG spec, but other volumes are under spec.volumes. this means we have flags like  `rootVolumeRetainOnTermination`... I realize this is only partially related to your PR, since you're building onto what i'm seeing as inherently inconsistent.
2. It looks like the `retainOnTermination` flag on a non-root volume is tied to the `EbsDeleteOnTermination` setting in the LaunchConfig. so we end up setting the opposite of what we want to do... sort of? 

I think it may make more sense (at least to me) to be configuring the option `deleteOnTermination`-- just based on the way it gets translated into the LC. if you were to go that way, I think that we should adopt the same perspective for the root volume... and someday move that into spec.volumes. 

all that said, your feature works. I just want to make sure that we're doing the best we can to maintain/improve UX.  further to that, could you please add a blurb to the docs explaining how these flags work? that would be super helpful.

many thanks for your time and submission! and please feel free to educate me where i'm wrong :D

**Comment by tioxy** [comment]
@geojaz Pulled the `make ci` change to a different PR :+1: 

You are right about intentions and UX, i feel the same tbh.

### Suggestion
I think we need to stick with one of the following nomenclatures:
- deleteOnTermination
- retainOnTermination


Cool information to consider about cloud providers naming this strategy:
- AWS: [`DeleteOnTermination`](https://godoc.org/github.com/aws/aws-sdk-go/service/ec2#EbsBlockDevice)
- GCP: [`AutoDelete`](https://godoc.org/google.golang.org/api/compute/v1#AttachedDisk)

Also, in the previous PR, there was a discussion starting from the following comment about the nomenclature https://github.com/kubernetes/kops/pull/6373#discussion_r266287230

Honestly, it feels like a better experience using `deleteOnTermination`. Our mainly supported cloud providers use *Delete* instead of *Retain* when naming this kind of policy for their disk APIs.

I also think moving `RootVolumeTermination` ➡️ `RootVolumeDeleteOnTermination` would feel better in `kops update cluster` as it would match the InstaceGroup API better.

### Example

```yaml
 kind: InstanceGroup
 metadata:
   name: master-us-east-1a
 spec:
   role: Master
   rootVolumeDeleteOnTermination: false
   volumes:
   - device: /dev/xvdd
     encrypted: true
     size: 20
     type: gp2
```

```yaml
 kind: InstanceGroup
 metadata:
   name: nodes
 spec:
   role: Node
   volumes:
   - device: /dev/xvdd
     encrypted: true
     deleteOnTermination: false
     size: 20
     type: gp2
```

```go
  LaunchConfiguration/master-us-east-1a.masters.kops.tioxy.com
  	BlockDeviceMappings         	[{"DeviceName":"/dev/xvdd","EbsDeleteOnTermination":true,"EbsEncrypted":true,"EbsVolumeIops":null,"EbsVolumeSize":20,"EbsVolumeType":"gp2","VirtualName":null}]
  	RootVolumeSize              	64
  	RootVolumeDeleteOnTermination	false
  	RootVolumeType              	gp2	

  LaunchConfiguration/nodes.kops.tioxy.com
  	BlockDeviceMappings         	[{"DeviceName":"/dev/xvdd","EbsDeleteOnTermination":false,"EbsEncrypted":true,"EbsVolumeIops":null,"EbsVolumeSize":20,"EbsVolumeType":"gp2","VirtualName":null}]
  	RootVolumeSize              	128
  	RootVolumeDeleteOnTermination	true
  	RootVolumeType              	gp2   
```

### Logic difference

The following example is the logic behind `deleteOnTermination` vs `retainOnTermination`
```go
// deleteOnTermination

deleteRootVolume := DefaultVolumeDeleteOnTermination
if ig.Spec.RootVolumeDeleteOnTermination != nil {
    deleteRootVolume = fi.BoolValue(ig.Spec.RootVolumeRetainOnTermination)
}
```

```go
// retainOnTermination

deleteRootVolume := DefaultVolumeDeleteOnTermination
retainRootVolume := fi.BoolValue(ig.Spec.RootVolumeRetainOnTermination)
if retainRootVolume {
	deleteRootVolume = false
}
```

> It even feels more readable and cleaner the `deleteOnTermination`


### Conclusion
I can rewrite without any problems and also add the documentation, are we up to change it like this?

**Comment by joshbranham** [comment]
I would agree that using the following nomenclature makes the most sense: `deleteOnTermination`, with `rootVolume` in front etc as you have described, thanks for continuing to work on this as we deliberate :)

**Comment by rifelpet** [comment]
I agree it looks good! Can we mention the default value of `true` in the API field comments so that it shows up in the API documentation?

And I noticed this will only apply to LaunchConfigurations and not LaunchTemplates:

https://github.com/kubernetes/kops/blob/e1f7d31c823a4ad3dff2c8767c2f7c8d361738d7/upup/pkg/fi/cloudup/awstasks/launchtemplate.go#L103-L108

I think we can address that in a followup PR but it might be worth mentioning in the API field comments that it currently only applies to LaunchConfigurations.

**Comment by tioxy** [comment]
@rifelpet Pushed the default/launch configuration api docs with new codegen

Thank you all for such great reviews :+1:

**Comment by rifelpet** [comment]
One last concern came to mind: if volumes are not deleted by the ASG, does a `kops delete cluster` delete them? We should determine whether it does or not and clearly document that as well. I could see either behavior being desired, so it's important to make sure users are aware of whichever one it is.

File: k8s/crds/kops.k8s.io_instancegroups.yaml [hunk 1]
```diff
@@ -624,6 +624,11 @@ spec:
               description: 'Type determines the role of instances in this group: masters
                 or nodes'
               type: string
+            rootVolumeDeleteOnTermination:
+              description: RootVolumeDeleteOnTermination states if the root volume
+                will be deleted after instance termination. Default true. Only supported
+                by Launch Configuration.
```
[the remark this answers is not in the document]
Comment (tioxy) [reply k8s/crds/kops.k8s.io_instancegroups.yaml]: Actually nice suggestion, straightforward

**Comment by tioxy** [comment]
@geojaz Updated API field docs with your suggestion, way more human readable and straightforward :sweat_smile:  

@rifelpet Volumes are retained after cluster deletion. Just added this to API field docs as well.

**Pull request merged**